reward/supply_manager.go

```diff
@@ -0,0 +1,377 @@
+	// The head block can be obsolete by the time catchup finished, so the big-step can end up being a bit short.
+	// Repeat until the difference is close enough so that the headNum stays the same after one iteration.
+	for lastNum < headNum {
+		logger.Info("Total supply big step catchup", "last", lastNum, "head", headNum, "minted", lastAccReward.Minted.String(), "burntFee", lastAccReward.BurntFee.String())
+
+		accReward, err := sm.accumulateReward(lastNum, headNum, lastAccReward)
+		if err != nil {
+			if err != errSupplyManagerQuit {
+				logger.Error("Total supply accumulate failed", "from", lastNum, "to", headNum, "err", err)
+			}
+			return
+		}
+
+		lastNum = headNum
+		lastAccReward = accReward
+		headNum = sm.chain.CurrentBlock().NumberU64()
+	}
+	logger.Info("Total supply big step catchup done", "last", lastNum, "minted", lastAccReward.Minted.String(), "burntFee", lastAccReward.BurntFee.String())
+
+	// Subscribe to chain head events and accumulate on demand.
+	sm.chainHeadSub = sm.chain.SubscribeChainHeadEvent(sm.chainHeadChan)
+	for {
+		select {
+		case <-sm.quitCh:
+			return
+		case head := <-sm.chainHeadChan:
+			headNum = head.Block.NumberU64()
+
+			supply, err := sm.accumulateReward(lastNum, headNum, lastAccReward)
+			if err != nil {
+				if err != errSupplyManagerQuit {
+					logger.Error("Total supply accumulate failed", "from", lastNum, "to", headNum, "err", err)
+				}
+				return
+			}
+
+			lastNum = headNum
+			lastAccReward = supply
+		}
+	}
+}
+
+// totalSupplyFromState calculates the ground truth total supply by iterating over all accounts.
+// This is extremely inefficient and should only be used for the genesis block and testing.
+func (sm *supplyManager) totalSupplyFromState(num uint64) (*big.Int, error) {
+	header := sm.chain.GetHeaderByNumber(num)
+	if header == nil {
+		return nil, errors.New("header not found")
+	}
+	stateDB, err := sm.chain.StateAt(header.Root)
+	if err != nil {
+		return nil, err
+	}
+	dump := stateDB.RawDump() // Extremely inefficient but okay for genesis block.
+
+	totalSupply := new(big.Int)
+	for _, account := range dump.Accounts {
+		balance, ok := new(big.Int).SetString(account.Balance, 10)
+		if !ok {
+			return nil, errors.New("malformed state dump")
+		}
+		totalSupply.Add(totalSupply, balance)
+	}
+	return totalSupply, nil
+}
+
+// accumulateReward calculates the total supply from the last block to the current block.
+// Given supply at `from` is `fromSupply`, calculate the supply until `to`, inclusive.
+func (sm *supplyManager) accumulateReward(from, to uint64, fromAcc *database.AccReward) (*database.AccReward, error) {
+	accReward := fromAcc.Copy() // make a copy because we're updating it in-place.
+
+	for num := from + 1; num <= to; num++ {
+		// Abort upon quit signal
+		if atomic.LoadUint32(&sm.quit) != 0 {
+			return nil, errSupplyManagerQuit
+		}
+
+		// Accumulate one block
+		var (
+			header    = sm.chain.GetHeaderByNumber(num)
```

Comment:
Doesn't it need to stop the accumulation if it meets `header == nil`?

Reply:
I haven't met such a case. If `header == nil` it should be an error.
The accumulateReward is called only if we know that there's a header at the `to` block.

Reply:
Ok, I think it's redundant. Thanks.